### PR TITLE
Add more decoders for cells

### PR DESCRIPTION
**Pull request by ybasket**
Plenty of standard cell decoders (including literals on 2.13), some helpful methods on `CellDecoder` itself (naming inspired by circe), simple tests.

### CODE REVIEW

File: csv/src/fs2/data/csv/CellDecoder.scala [hunk 1]
```diff
@@ -15,21 +15,63 @@
...
+  /**
+   * Fail-over. If this decoder fails, try the supplied other decoder.
+   * @param cd the fail-over decoder
+   * @tparam TT the return type
+   * @return a decoder combining this and the other decoder
+   */
+  def or[TT >: T](cd: => CellDecoder[TT]): CellDecoder[TT] = s =>
+    apply(s) match {
+      case Left(_) => cd(s)
+      case Right(value) => value.asRight
```
Comment (satabin) [inline csv/src/fs2/data/csv/CellDecoder.scala]:
```suggestion
      case r @ Right(_) => value.leftCast[DecoderError]
```

File: csv/src/fs2/data/csv/CellDecoder.scala [hunk 2]
```diff
@@ -53,25 +95,94 @@ object CellDecoder {
...
+    Try(BigInt(s)).toEither.leftMap(new DecoderError(s"unable to decode '$s' as a BigInt", _))
   implicit val stringDecoder: CellDecoder[String] = s => Right(s)
+  implicit val charArrayDecoder: CellDecoder[Array[Char]] = s => Right(s.toCharArray)
+
+  // Containers
+  implicit def decoderResultDecoder[T](implicit ev: CellDecoder[T]): CellDecoder[DecoderResult[T]] = s =>
+    ev(s).asRight
+  implicit def rawOrResultDecoder[T](implicit ev: CellDecoder[T]): CellDecoder[Either[String, T]] = s =>
+    ev(s).leftMap(_ => s).asRight
+  implicit def optionDecoder[T](implicit ev: CellDecoder[T]): CellDecoder[Option[T]] = s =>
```
Comment (satabin) [inline csv/src/fs2/data/csv/CellDecoder.scala]: This is dangerous and might conflict with the `[Csv]RowDecoder` handling of optional columns, I think. Have you checked?

File: csv/src/fs2/data/csv/CellDecoder.scala [hunk 3]
```diff
@@ -53,25 +95,94 @@ object CellDecoder {
...
+
+  private def javaTimeDecoder[T](name: String, t: String => T): CellDecoder[T] = s =>
+    Either.catchOnly[DateTimeParseException](t(s))
+      .leftMap(t => new DecoderError(s"Couldn't parse $name: ${t.getMessage}", t))
+}
 
+trait CellDecoderInstances1 {
+  implicit val durationDecoder: CellDecoder[Duration] = s =>
+    Either.catchNonFatal(Duration(s))
+      .leftMap(t => new DecoderError(s"Couldn't parse Duration: ${t.getMessage}", t))
```
Comment (satabin) [inline csv/src/fs2/data/csv/CellDecoder.scala]: No need to repeat the `t` message in the wrapping error. Also convention I used in exception messages is to start with lower case.

File: csv/src/fs2/data/csv/CellDecoder.scala [hunk 4]
```diff
@@ -53,25 +95,94 @@ object CellDecoder {
...
+  implicit val localDateDecoder: CellDecoder[LocalDate] = javaTimeDecoder("LocalDate", LocalDate.parse)
+  implicit val localDateTimeDecoder: CellDecoder[LocalDateTime] = javaTimeDecoder("LocalDateTime", LocalDateTime.parse)
+  implicit val localTimeDecoder: CellDecoder[LocalTime] = javaTimeDecoder("LocalTime", LocalTime.parse)
+  implicit val offsetDateTimeDecoder: CellDecoder[OffsetDateTime] = javaTimeDecoder("OffsetDateTime", OffsetDateTime.parse)
+  implicit val offsetTimeDecoder: CellDecoder[OffsetTime] = javaTimeDecoder("OffsetTime", OffsetTime.parse)
+  implicit val zonedDateTimeDecoder: CellDecoder[ZonedDateTime] = javaTimeDecoder("ZonedDateTime", ZonedDateTime.parse)
+
+  private def javaTimeDecoder[T](name: String, t: String => T): CellDecoder[T] = s =>
+    Either.catchOnly[DateTimeParseException](t(s))
+      .leftMap(t => new DecoderError(s"Couldn't parse $name: ${t.getMessage}", t))
```
Comment (satabin) [inline csv/src/fs2/data/csv/CellDecoder.scala]: No need to repeat the t message in the wrapping error. Also convention I used in exception messages is to start with lower case.

File: csv/src-2.13/fs2/data/csv/LiteralCellDecoders.scala [hunk 5]
```diff
@@ -0,0 +1,69 @@
+package fs2.data.csv
+
+trait LiteralCellDecoders {
+
+  private[this] abstract class LiteralCellDecoder[A, L <: A](decodeA: CellDecoder[A], L: ValueOf[L]) extends CellDecoder[L] {
+    protected[this] def check(a: A): Boolean
+    protected[this] def message: String
+
+    final def apply(s: String): DecoderResult[L] = decodeA(s) match {
+      case r @ Right(value) if check(value) => r.asInstanceOf[DecoderResult[L]]
```
[the remark this answers is not in the document]
Comment (ybasket) [reply csv/src-2.13/fs2/data/csv/LiteralCellDecoders.scala]: Nope, it's not casting the left side, it's narrowing the right side.

[the remark this answers is not in the document]
Comment (satabin) [reply csv/src-2.13/fs2/data/csv/LiteralCellDecoders.scala]: Arf, ok.

File: build.sc [hunk 6]
```diff
@@ -20,7 +20,7 @@ val fs2DataUrl = "https://github.com/satabin/fs2-data"
 
 val fs2DataDeveloper = Developer("satabin", "Lucas Satabin", "https://github.com/satabin")
 
-trait Fs2DataModule extends ScalaModule with ScalafmtModule {
+trait Fs2DataModule extends CrossScalaModule with ScalafmtModule {
```
[the remark this answers is not in the document]
Comment (ybasket) [reply build.sc]: Not here, moved it down to more specific modules.

File: build.sc [hunk 7]
```diff
@@ -114,11 +115,13 @@ class JsonModule(val crossScalaVersion: String) extends Fs2DataModule with Cross
...
-      def moduleDeps = Seq(circe)
-    }
+  object test extends Fs2DataTests {
+    def moduleDeps = Seq(circe(outer.crossScalaVersion))
+  }
+
+  object circe extends Cross[CirceModule](scala212, scala213)
 
-  object circe extends Fs2DataModule with PublishModule {
+  class CirceModule(val crossScalaVersion: String) extends Fs2DataModule with CrossScalaModule with PublishModule {
```
Comment (satabin) [inline build.sc]: It should not be necessary, as this module is in a cross module, it will be published for all cross versions.

Comment (ybasket) [reply build.sc]: True, reverted. Had some problems when setting up the cross sources for the literals and that change seemed to fix them, not sure what was going on.

File: build.sc [hunk 8]
```diff
@@ -72,6 +72,7 @@ class CsvModule(val crossScalaVersion: String) extends Fs2DataModule with CrossS
 
   object generic extends Fs2DataModule with PublishModule {
     def scalaVersion = outer.scalaVersion
+    def crossScalaVersion = outer.crossScalaVersion
```
Comment (satabin) [inline build.sc]: So, this is not needed anymore :)

**Pull request merged**